### PR TITLE
Tcp deadlock fix

### PR DESCRIPTION
**Pull request by bufrr**
Fix deadlock when udp connection block tcp, and using `SetLinger(0)` to avoid to many CLOSE_WAIT / TIME_WAIT issue

### CODE REVIEW

File: entry.go [hunk 1]
```diff
@@ -426,6 +428,13 @@ func (te *TunaEntry) listenTCP(ip net.IP, ports []uint32) ([]uint32, error) {
 		go func() {
 			for {
 				conn, err := listener.Accept()
+				if c, ok := conn.(*net.TCPConn); ok {
+					err := c.SetLinger(0)
+					if err != nil {
+						log.Println("Couldn't set linger:", err)
+						return
```
Comment (yilunzhang) [inline entry.go:435]: Why `return` here but `continue` in reverse (L704)?

File: entry.go [hunk 2]
```diff
@@ -657,17 +649,11 @@ func StartReverse(config *EntryConfiguration, wallet *nkn.Wallet) error {
...
-				keyToAddr[k] = from.String()
-				keyToAddrLock.Unlock()
+				addrToKey.Store(from.String(), k)
+				keyToAddr.Store(k, from.String())
 
-				udpReadyLock.RLock()
-				closeChan(udpReady[k])
-				udpReadyLock.RUnlock()
+				c, _ := udpReady.Load(k)
+				closeChan(c.(chan struct{}))
```
Comment (yilunzhang) [inline entry.go:656]: should only do type conversion if load success

File: entry.go [hunk 3]
```diff
@@ -795,17 +781,15 @@ func StartReverse(config *EntryConfiguration, wallet *nkn.Wallet) error {
...
 
 					if len(metadata.ServiceUdp) > 0 {
-						tcpReadyLock.RLock()
-						<-tcpReady[connKey]
-						tcpReadyLock.RUnlock()
+						tr, _ := tcpReady.Load(connKey)
+						<-tr.(chan struct{})
+
+						ur, _ := udpReady.Load(connKey)
+						<-ur.(chan struct{})
```
Comment (yilunzhang) [inline entry.go:788]: should only do type conversion if load success

File: entry.go [hunk 4]
```diff
@@ -795,17 +781,15 @@ func StartReverse(config *EntryConfiguration, wallet *nkn.Wallet) error {
...
 
-						udpReadyLock.RLock()
-						<-udpReady[connKey]
-						udpReadyLock.RUnlock()
+						addr, _ := keyToAddr.Load(connKey)
+						clientAddr := addr.(string)
```
Comment (yilunzhang) [inline entry.go:791]: should only do type conversion if load success

File: entry.go [hunk 5]
```diff
@@ -795,28 +787,32 @@ func StartReverse(config *EntryConfiguration, wallet *nkn.Wallet) error {
...
-						go func() {
+							addr, ok := keyToAddr.Load(connKey)
+							if !ok {
+								return
+							}
+							clientAddr := addr.(string)
+
+							ip, portStr, err := net.SplitHostPort(clientAddr)
+							if err != nil {
+								log.Printf("parse host error: %v\n", err)
```
Comment (yilunzhang) [inline entry.go:811]: return?

File: entry.go [hunk 6]
```diff
@@ -828,22 +824,17 @@ func StartReverse(config *EntryConfiguration, wallet *nkn.Wallet) error {
...
 									}
-									atomic.AddUint64(&te.Common.reverseBytesExitToEntry[key][data[2]], uint64(n))
+									atomic.AddUint64(&te.Common.reverseBytesExitToEntry[key.(string)][data[2]], uint64(n))
 								case <-te.udpCloseChan:
 									return
 								}
 							}
 						}()
-
-						te.SetServerUDPReadChan(udpReadChan)
```
Comment (yilunzhang) [inline entry.go:845]: Why remove these?

Comment (bufrr) [reply entry.go:845]: > Why remove these?

Both init within `NewCommon` already

**Review by yilunzhang — APPROVED** [review]
LGTM

**Pull request merged**